### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [3]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [4]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [5]:
#Total Player    #Series.nunique(dropna=True) Return number of unique elements in the object.

total_players = purchase_data["SN"].nunique()
tot_players_dict={"Total Players":[total_players]}
tot_plyr_df = pd.DataFrame.from_dict(tot_players_dict, orient='columns')
tot_plyr_df



Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [6]:
#number of unique items
unique = purchase_data["Item ID"].value_counts().count()

#need to figure out how to output as $0.00
#average purchase amount
avg = purchase_data["Price"].mean()

#Total Number of Purchases
total_purch = purchase_data["Purchase ID"].value_counts().count()

#need to figure out how to output as $0.00
#Total Revenue
Total_Revenue = purchase_data["Price"].sum()

p_dict = {"Number of Unique Items":[unique],
        "Average Price":[avg],
          "Number of Purchases":[total_purch],
          "Total Revenue":[Total_Revenue]}
p_df = pd.DataFrame.from_dict(p_dict, orient='columns')
p_df["Average Price"]= p_df["Average Price"].map("${:,.2f}".format)
p_df["Total Revenue"]= p_df["Total Revenue"].map("${:,.2f}".format)
p_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [7]:
unique_users_df = purchase_data.drop_duplicates(["SN"])
gender_group_df = unique_users_df.groupby(['Gender'])
gender_dict = {"Percentage of Players": gender_group_df["Gender"].count()/total_players*100,
               "Total Count": gender_group_df["SN"].count()}

gender_df = pd.DataFrame.from_dict(gender_dict,orient='columns')
gender_df = gender_df.sort_values("Total Count", ascending=False)


gender_df ["Percentage of Players"]= gender_df["Percentage of Players"].map("{:.2f}%".format)
gender_df.head()

Percentage of Players  Total Count
Gender                                                  
Male                                 84.03%          484
Female                               14.06%           81
Other / Non-Disclosed                 1.91%           11


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [8]:
gender_df = purchase_data.groupby(['Gender'])
purchase_dict = {"Purchase Count": gender_df["Purchase ID"].count(),
                 "Average Purchase Price": gender_df["Price"].mean(),
                "Total Purchase Value": gender_df["Price"].sum(),
                "Average Total Purchase per Person": gender_df["Price"].sum()/ gender_df["SN"].nunique()
                }
                 
purchase_df = pd.DataFrame.from_dict(purchase_dict,orient='columns')    

purchase_df["Average Purchase Price"]= purchase_df["Average Purchase Price"].map("${:,.2f}".format)
purchase_df["Total Purchase Value"]= purchase_df["Total Purchase Value"].map("${:,.2f}".format)
purchase_df["Average Total Purchase per Person"]= purchase_df["Average Total Purchase per Person"].map("${:,.2f}".format)

purchase_df



Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Average Total Purchase per Person  
Gender                                                                        
Female                             $361.94                             $4.47  
Male                             $1,967.64                             $4.07  
Other / Non-Disclosed               $50.19                             $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [9]:
#### Create the bins in which Data will be held
  
bins = [0, 9,14,19,24,29,34,39,100]

# Create the names for the four bins
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]



In [10]:
purchase_data["Age_G"] = pd.cut(purchase_data["Age"], bins, labels=group_names)

In [11]:
df = purchase_data.groupby("Age_G")
age_dict = {"Total Count":df["SN"].nunique(),
            "Percentage of Players":df["SN"].nunique()/total_players*100
           }
age_df = pd.DataFrame.from_dict(age_dict,orient='columns')
age_df["Percentage of Players"] = age_df["Percentage of Players"].map("{:.2f}%".format)

age_df

Total Count Percentage of Players
Age_G                                   
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [12]:
age_dict2 = {"Purchase Count":df["Purchase ID"].nunique(),
            "Average Purchase Price":df["Price"].mean(),
            "Total Purchase Value": df["Price"].sum(),
            "Average Total Purchase Per Person": df["Price"].sum()/df["SN"].nunique()
           }
age_df2 = pd.DataFrame.from_dict(age_dict2,orient='columns')
age_df2["Average Purchase Price"]= age_df2["Average Purchase Price"].map("${:,.2f}".format)
age_df2["Total Purchase Value"]= age_df2["Total Purchase Value"].map("${:,.2f}".format)
age_df2["Average Total Purchase Per Person"]= age_df2["Average Total Purchase Per Person"].map("${:,.2f}".format)
age_df2

Purchase Count Average Purchase Price Total Purchase Value  \
Age_G                                                               
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Average Total Purchase Per Person  
Age_G                                    
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [13]:
user_df = purchase_data.groupby("SN")
user_dict = {"Purchase Count":user_df["Purchase ID"].count(),
             "Average Purchase Price": user_df["Price"].mean(),
             "Total Purchase Value":user_df["Price"].sum() }
user_df2 = pd.DataFrame.from_dict(user_dict,orient='columns')
user_df2 = user_df2.sort_values("Total Purchase Value", ascending=False)

user_df2["Average Purchase Price"]= user_df2["Average Purchase Price"].map("${:,.2f}".format)
user_df2["Total Purchase Value"]= user_df2["Total Purchase Value"].map("${:,.2f}".format)
user_df2.head()


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [15]:
reduced_purchase_data_df = purchase_data.loc[:, ["Item ID", "Item Name", "Price"]]

grouped_purchase_data_df = reduced_purchase_data_df.groupby(["Item ID", "Item Name"])

purch_dict = {"Purchase Count":grouped_purchase_data_df["Item ID"].count(),
             "Item Price": grouped_purchase_data_df["Price"].mean(),
             "Total Purchase Value":grouped_purchase_data_df["Price"].sum() }
purch_df2 = pd.DataFrame.from_dict(purch_dict,orient='columns')
purch_df2 = purch_df2.sort_values("Purchase Count", ascending=False)

purch_df2["Item Price"]= purch_df2["Item Price"].map("${:,.2f}".format)
purch_df2["Total Purchase Value"]= purch_df2["Total Purchase Value"].map("${:,.2f}".format)

purch_df2.head()



Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [202]:
purch_df3 = pd.DataFrame.from_dict(purch_dict,orient='columns')
purch_df3 = purch_df3.sort_values("Total Purchase Value", ascending=False)

purch_df3["Item Price"]= purch_df3["Item Price"].map("${:,.2f}".format)
purch_df3["Total Purchase Value"]= purch_df3["Total Purchase Value"].map("${:,.2f}".format)

purch_df3.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

Three Observable Trends 

1) About 45% of the players are between the ages of 20 and 24.
2) Players between the ages of 35 to 39 spend the most money per user but are only about 5% of the players.
3) Oathbreaker, Last Home of the Breaking Storm is the most purchased content package and also brings in the most revenue.